ssl/t1_lib.c

```diff
@@ -417,6 +417,313 @@ int ssl_load_groups(SSL_CTX *ctx)
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MIN_DTLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->mindtls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MAX_DTLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->maxdtls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+    /*
+     * Now check that the algorithm is actually usable for our property query
+     * string. Regardless of the result we still return success because we have
+     * successfully processed this signature, even though we may decide not to
+     * use it.
+     */
+    ret = 1;
+    ERR_set_mark();
+    keymgmt = EVP_KEYMGMT_fetch(ctx->libctx, sinf->keytype?sinf->keytype:
+		    sinf->sig_name?sinf->sig_name:sinf->sigalg_name, ctx->propq);
+    if (keymgmt != NULL) {
+        /*
+         * We have successfully fetched the algorithm - however if the provider
+         * doesn't match this one then we ignore it.
+         *
+         * Note: We're cheating a little here. Technically if the same algorithm
+         * is available from more than one provider then it is undefined which
+         * implementation you will get back. Theoretically this could be
+         * different every time...we assume here that you'll always get the
+         * same one back if you repeat the exact same fetch. Is this a reasonable
+         * assumption to make (in which case perhaps we should document this
+         * behaviour)?
+         */
+        if (EVP_KEYMGMT_get0_provider(keymgmt) == provider) {
+            /* We have a match - so we could use this signature */
+            /* Check proper object registration first, though */
+            /* don't care about return value as this may have been
+             * done within providers or previous calls to
+             * add_provider_sigalgs
+             */
+            OBJ_create(sinf->sigalg_oid, sinf->sigalg_name, NULL);
+            if (sinf->sig_name != NULL)
+                OBJ_create(sinf->sig_oid, sinf->sig_name, NULL);
+            if (sinf->keytype != NULL)
+                OBJ_create(sinf->keytype_oid, sinf->keytype, NULL);
+	    if (sinf->hash_name != NULL) {
+                OBJ_create(sinf->hash_oid, sinf->hash_name, NULL);
+                if (sinf->sig_name != NULL)
+                    OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
+                                  OBJ_txt2nid(sinf->hash_name),
+                                  OBJ_txt2nid(sinf->sig_name));
+            } else {
+                OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
+                              NID_undef,
+                              (sinf->sig_name == NULL)?NID_undef:
+                                             OBJ_txt2nid(sinf->sig_name));
```

Comment:
This doesn't make sense when `sinf->sig_name == NULL` (that will return an error).  I suggest this instead:

```suggestion
                OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
                              NID_undef,
                              (sinf->sig_name != NULL)
                              ? OBJ_txt2nid(sinf->sig_name)
                              : OBJ_txt2nid(sinf->sigalg_name));
```

Reply:
Gah!  I'm actually undecided on this...  it could also have been this:

```suggestion
                OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
                              NID_undef,
                              (sinf->sig_name != NULL)
                              ? OBJ_txt2nid(sinf->sig_name)
                              : OBJ_txt2nid(sinf->keytype));
```

This would make sense in that the pure signature algorithm name is often the same as the key type (at least classically, with RSA and DSA...  ECC kinda threw that assumption out the Windows), so perhaps the better choice after all...

Reply:
Unsure what to do now: Leave as-is or change to the new proposal? For some reason the button "Commit suggestion" is greyed out/inaccessible to me.

Reply:
I put together a solution that I believe is better.

Reply:
@levitte Implementing this change suggestion would break the code: `keytype` is optional and would break if `keytype` is not set (actually, _does_ break in testing :) This code would be OK: 
```
                OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
                              NID_undef,
                              (sinf->sig_name != NULL)
                              ? OBJ_txt2nid(sinf->sig_name)
                              : (sinf->keytype != NULL)?
                                   OBJ_txt2nid(sinf->keytype):
                                   OBJ_txt2nid(sinf->sigalg_name));
```
--> OK to add?

Reply:
OK to close this conversation and continue with your proposal below?